### ISSUE TITLE
[no-misused-promises]Is there a breaking change between 5.13.0 and 5.12.1?

### ISSUE THREAD

**Issue by xtx1130** [typescript-eslint/typescript-eslint#4616]
- [x] I have tried restarting my IDE and the issue persists.
- [ ] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-misused-promises": ["error"]
  }
}
```

case: https://github.com/fastify/fastify/blob/main/test/types/hooks.test-d.ts#L142-L146
```TS
// your repro code case
server.addHook('onRequest', async function (request, reply) {
  expectType<FastifyInstance>(this)
  expectType<FastifyRequest>(request)
  expectType<FastifyReply>(reply)
})
```

**Expected Result**
When I `npm run lint` with `typescript-eslint/eslint-plugin@5.12.1` in [fastify](https://github.com/fastify/fastify), there is no error occured:
<img width="600" alt="image" src="https://user-images.githubusercontent.com/5475069/156150561-e39d38fc-d2cf-4346-9ddd-34bd697e2769.png">

**Actual Result**
 it will throw error in  `typescript-eslint/eslint-plugin@5.13.0`:
<img width="600" alt="image" src="https://user-images.githubusercontent.com/5475069/156150695-a25c5543-9bdc-414c-aae9-f0c561073de1.png">

**Additional Info**
none

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.13.0` |
| `@typescript-eslint/parser`        | `5.13.0` |
| `TypeScript`                       | `4.6.2` |
| `ESLint`                           | `8.10.0` |
| `node`                             | `16.13.1` |

**Comment by xtx1130** [comment]
I think it is the same question with https://github.com/typescript-eslint/typescript-eslint/issues/4609
